hypothesis-python/src/hypothesis/core.py

```diff
@@ -1248,6 +1250,18 @@ def wrapped_test(*arguments, **kwargs):
                     "to ensure that each example is run in a separate "
                     "database transaction."
                 )
+            if settings.database is not None and "self" in kwargs:
+                nonlocal prev_self
+                if prev_self is None:
+                    prev_self = kwargs["self"]
+                elif prev_self != kwargs["self"]:
```

Comment:
Looking this up in `kwargs` is unsafe - there might be no kwargs, under a different name, etc.  Happily we've already got logic to sort all that out and `runner = stuff.selfy` a few lines up will be our class instance if there is one, or None.  Note we want a non-None sentinel so that we can still tell if someone does something _really_ weird...

Reply:
`stuff.selfy` is too indiscriminate, we don't want to trigger on f.x. different values of `implementation` here.

https://github.com/HypothesisWorks/hypothesis/blob/621fed2a067b379b24ccca031a80160eee9ea481/hypothesis-python/tests/cover/test_cache_implementation.py#L83-L90